New Checker Openjpeg checker

chaitanyamogal opened the pull request:
Added new checker openjpeg. As the openjpeg package is shifted to version 2 and the test file is available by the name openjpeg2. #976 
I am not able to figure out why the tests are not passing. please help.

terriko commented on cve_bin_tool/checkers/openjpeg2.py:
```diff
@@ -0,0 +1,16 @@
+    CONTAINS_PATTERNS = []
+    FILENAME_PATTERNS = [r"libopenjp2\.so", r"openjpeg2"]
+    VERSION_PATTERNS = [r"openjpeg2 ([0-9]+\.[0-9]+\.[0-9]+)"]
+    VENDOR_PRODUCT = [("openjpeg2")]
```
VENDOR_PRODUCT has to be two items (a vendor and a product) in order to work.

If you search for openjpeg in something like cvedetails, it looks like it's not openjpeg2 in the database:

https://www.cvedetails.com/google-search-results.php?q=openjpeg&sa=Search

so you'll need to fix that.

terriko commented:
Also, you're even failing the black check at the moment, so you should make sure to fix that as well!

chaitanyamogal commented:
> Also, you're even failing the black check at the moment, so you should make sure to fix that as well!

I have fixed the black check and also added product and vendor but still. But still, tests are failing.

terriko commented:
Have you looked at the error messages?

The one on the short test looks like this (I've omitted some of the code it shows before the error)

```
>           assert product in list_products
E           AssertionError: assert 'openjpeg2' in {'openjpeg'}

test/test_scanner.py:128: AssertionError
------------------------------ Captured log call -------------------------------
INFO     cve_bin_tool.VersionScanner:version_scanner.py:163 /tmp/mapping-test-pv333vsq/tmp0vzh4pay-openjpeg2-2.3.1.out is openjpeg2 2.3.1
=============================== warnings summary ===============================
../../../../../opt/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/setuptools/depends.py:2
  /opt/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/setuptools/depends.py:2: DeprecationWarning:
  
  the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED test/test_scanner.py::TestScanner::test_version_mapping[openjpeg2-2.3.1-version_strings46]
======= 1 failed, 183 passed, 159 skipped, 1 warning in 65.81s (0:01:05) =======
```

You should be seeing the same one when you run the short tests in your console, but you can also view it on this pull request under the "checks" tab (useful when it works for you locally but doesn't work in CI, which sometimes happens due to differences in the local and CI environments)

So there's a warning in there about setuptools that isn't  your problem which may have confused you when you looked, but you can ignore this warning (you can tell that you can ignore it because it lists the file it occurred in, and it's part of setuptools not cve-bin-tool)

But if you look at the top of what I cut and pasted, you can see this error:

```
>           assert product in list_products
E           AssertionError: assert 'openjpeg2' in {'openjpeg'}
```

We use assertions to indicate things that should be true when the test runs.  So your test asserted that the code should find `openjpeg2` but what it actually found was `openjpeg` -- if you change the test to expect to find openjpeg, it should start passing.

Try on your own for the other failing long test!  What is it looking for and what is it finding?

terriko commented:
Oh, and if you're still having trouble with grepping and whatnot on windows, remember that you can put the output from strings into your favourite text editor and search for "openjpeg" to see what good candidate signatures might look like (since it looks like it's not finding openjpeg at the moment, which means your signature isn't good).

pdxjohnny commented:
I don't this the `VERSION_PATTERNS` is correct. What lead you do this version pattern?

What would help is trying to find an `.rpm` (or more than one) or a package which contains openjpeg so that you can add a long test.

I usually search on pkgs.org: https://pkgs.org/search/?q=openjpeg and extract using rpm2cpio and cpio or using [rpmfile](https://pypi.org/project/rpmfile/)

In the below example I've picked fedora 33's version of the 1.5.1 openjpeg package (they also have a version 2 package)

```console
$ curl -sfL https://download-ib01.fedoraproject.org/pub/fedora/linux/releases/33/Everything/x86_64/os/Packages/o/openjpeg-1.5.1-28.fc33.x86_64.rpm | python -m rpmfile -xv -
/tmp/tmp.xHYVmXC2B1/usr/bin/extract_j2k_from_mj2
/tmp/tmp.xHYVmXC2B1/usr/bin/frames_to_mj2
/tmp/tmp.xHYVmXC2B1/usr/bin/image_to_j2k
/tmp/tmp.xHYVmXC2B1/usr/bin/j2k_dump
/tmp/tmp.xHYVmXC2B1/usr/bin/j2k_to_image
/tmp/tmp.xHYVmXC2B1/usr/bin/mj2_to_frames
/tmp/tmp.xHYVmXC2B1/usr/bin/wrap_j2k_in_mj2
/tmp/tmp.xHYVmXC2B1/usr/lib/.build-id/12/f08f461c8cf7eaaa161f66ac374068e937aeac
/tmp/tmp.xHYVmXC2B1/usr/lib/.build-id/30/b147becf83536b9ecb59be3c902aa2be442c9d
/tmp/tmp.xHYVmXC2B1/usr/lib/.build-id/31/63969ccc2ccaf900c33e161fc6252f24dde34a
/tmp/tmp.xHYVmXC2B1/usr/lib/.build-id/6f/c1f46ffad20501cb58b0b3b03cd178f976b9f6
/tmp/tmp.xHYVmXC2B1/usr/lib/.build-id/7c/8a4a08788c0b2cffea5fa4b5c00161a8acf564
/tmp/tmp.xHYVmXC2B1/usr/lib/.build-id/b0/7c27f0b673710ebe17443fc23623d2b6e7f197
/tmp/tmp.xHYVmXC2B1/usr/lib/.build-id/da/a58edfbc4d6d89a560563691a4de8ff0f04088
/tmp/tmp.xHYVmXC2B1/usr/share/man/man1/image_to_j2k.1.gz
/tmp/tmp.xHYVmXC2B1/usr/share/man/man1/j2k_dump.1.gz
/tmp/tmp.xHYVmXC2B1/usr/share/man/man1/j2k_to_image.1.gz
```

I then look for binaries

```console
$ find . -exec file {} \; | grep executable, | tee >(sed -e 's/:.*//g' > executables.txt)
./usr/bin/wrap_j2k_in_mj2: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=3163969ccc2ccaf900c33e161fc6252f24dde34a, for GNU/Linux 3.2.0, stripped
./usr/bin/mj2_to_frames: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=b07c27f0b673710ebe17443fc23623d2b6e7f197, for GNU/Linux 3.2.0, stripped
./usr/bin/j2k_to_image: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=7c8a4a08788c0b2cffea5fa4b5c00161a8acf564, for GNU/Linux 3.2.0, stripped
./usr/bin/j2k_dump: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=12f08f461c8cf7eaaa161f66ac374068e937aeac, for GNU/Linux 3.2.0, stripped
./usr/bin/image_to_j2k: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=daa58edfbc4d6d89a560563691a4de8ff0f04088, for GNU/Linux 3.2.0, stripped
./usr/bin/frames_to_mj2: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=30b147becf83536b9ecb59be3c902aa2be442c9d, for GNU/Linux 3.2.0, stripped
./usr/bin/extract_j2k_from_mj2: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=6fc1f46ffad20501cb58b0b3b03cd178f976b9f6, for GNU/Linux 3.2.0, stripped
```

I run strings on those binaries and do a case insensitive search for the packages name (`grep -i`)

```console
$ strings $(cat executables.txt) | sort | uniq | grep -i openjpeg
/builddir/build/BUILD/openjpeg-1.5.1/applications/codec/convert.c
/builddir/build/BUILD/openjpeg-1.5.1/applications/common/color.c
/builddir/build/BUILD/openjpeg-1.5.1/applications/mj2/extract_j2k_from_mj2.c
/builddir/build/BUILD/openjpeg-1.5.1/applications/mj2/mj2_to_frames.c
/builddir/build/BUILD/openjpeg-1.5.1/applications/mj2/wrap_j2k_in_mj2.c
Created by OpenJPEG version 
libopenjpeg.so.1
OpenJPEG cannot encode raw components with bit depth higher than 16 bits.
# OpenJPEG-%s
#OpenJPEG-%s
```

I also look for the version number

```console
$ strings $(cat executables.txt) | sort | uniq | grep -i 1.5.1
>1.5.1
@1.5.1
1.5.1
/builddir/build/BUILD/openjpeg-1.5.1/applications/codec/convert.c
/builddir/build/BUILD/openjpeg-1.5.1/applications/common/color.c
/builddir/build/BUILD/openjpeg-1.5.1/applications/mj2/extract_j2k_from_mj2.c
/builddir/build/BUILD/openjpeg-1.5.1/applications/mj2/mj2_to_frames.c
/builddir/build/BUILD/openjpeg-1.5.1/applications/mj2/wrap_j2k_in_mj2.c
extract_j2k_from_mj2-1.5.1-28.fc33.x86_64.debug
frames_to_mj2-1.5.1-28.fc33.x86_64.debug
image_to_j2k-1.5.1-28.fc33.x86_64.debug
j2k_dump-1.5.1-28.fc33.x86_64.debug
j2k_to_image-1.5.1-28.fc33.x86_64.debug
mj2_to_frames-1.5.1-28.fc33.x86_64.debug
wrap_j2k_in_mj2-1.5.1-28.fc33.x86_64.debug
```

I'll leave the rest to you. Ping us if you need help! You may end up needing to overwrite the [`get_version()`](
https://github.com/intel/cve-bin-tool/blob/748ddb0092747f966b8213ed496d82d4944d1b40/cve_bin_tool/checkers/__init__.py#L120-L132) method for the checker, but that should be a last resort if you can't find a regex that works for `VERSION_PATTERNS`

chaitanyamogal commented:
hey, @pdxjohnny  and @terriko  please review the changes I made but still long test is failing. Is there a need to add a checker in test_checkers.py

pdxjohnny commented:
It looks like the long test you added is not passing

```

=================================== FAILURES ===================================
_ TestScanner.test_version_in_package[https://kojipkgs.fedoraproject.org/packages/openjpeg/1.5.1/28.fc33/x86_64/-openjpeg-1.5.1-28.fc33.x86_64.rpm-openjpeg-1.5.1] _
[gw1] linux -- Python 3.8.6 /opt/hostedtoolcache/Python/3.8.6/x64/bin/python

self = <test.test_scanner.TestScanner object at 0x7ffb8d10f1f0>
url = 'https://kojipkgs.fedoraproject.org/packages/openjpeg/1.5.1/28.fc33/x86_64/'
package_name = 'openjpeg-1.5.1-28.fc33.x86_64.rpm', product = 'openjpeg'
version = '1.5.1'

    @pytest.mark.parametrize(
        "url, package_name, product, version",
        (
            (d["url"], d["package_name"], d["product"], d["version"])
            for list_data in package_test_data
            for d in list_data
        ),
    )
    @unittest.skipUnless(LONG_TESTS() > 0, "Skipping long tests")
    def test_version_in_package(self, url, package_name, product, version):
        """Helper function to get a file (presumed to be a real copy
        of a library, probably from a Linux distribution) and run a
        scan on it.  Any test using this should likely be listed as a
        long test."""
        # get file
        tempfile = os.path.join(self.package_test_dir, package_name)
        download_file(url + package_name, tempfile)
        # new scanner for the new test.
        # self.scanner = VersionScanner(self.cve_scanner, should_extract=True)
        # run the tests
        list_products = set()
        list_versions = set()
        for scan_info in self.scanner.recursive_scan(tempfile):
            if scan_info:
                product_info, file_path = scan_info
                list_products.add(product_info.product)
                list_versions.add(product_info.version)
    
        # Make sure the product and version are in the results
>       assert product in list_products
E       AssertionError: assert 'openjpeg' in {'gcc', 'glibc'}

test/test_scanner.py:160: AssertionError
```

You probably need to add a [`get_version()`](
https://github.com/intel/cve-bin-tool/blob/748ddb0092747f966b8213ed496d82d4944d1b40/cve_bin_tool/checkers/__init__.py#L120-L132) method in your checker. Look through the code in the base class to understand what it's doing, and look through `regex_find()` in `utils.py`

chaitanyamogal commented:
Hey, @pdxjohnny  I didn't get it. Can you please explain in detail? Why the long tests are failing.

terriko commented:
Okay, once again, let's look at the errors:

```
>       assert product in list_products
E       AssertionError: assert 'openjpeg' in {'glibc'}

test/test_scanner.py:160: AssertionError
------------------------------ Captured log call -------------------------------
INFO     cve_bin_tool.VersionScanner:version_scanner.py:162 /tmp/package_test-fww1_tm6/openjpeg-1.5.1-18.el7.x86_64.rpm contains /usr/bin/mj2_to_frames contains glibc 2.15
INFO     cve_bin_tool.VersionScanner:version_scanner.py:162 /tmp/package_test-fww1_tm6/openjpeg-1.5.1-18.el7.x86_64.rpm contains /usr/bin/extract_j2k_from_mj2 contains glibc 2.15
INFO     cve_bin_tool.VersionScanner:version_scanner.py:162 /tmp/package_test-fww1_tm6/openjpeg-1.5.1-18.el7.x86_64.rpm contains /usr/bin/j2k_dump contains glibc 2.15
WARNING  cve_bin_tool.VersionScanner:version_scanner.py:154 openjpeg was detected with version UNKNOWN in file /tmp/package_test-fww1_tm6/openjpeg-1.5.1-18.el7.x86_64.rpm contains /usr/bin/j2k_dump
INFO     cve_bin_tool.VersionScanner:version_scanner.py:162 /tmp/package_test-fww1_tm6/openjpeg-1.5.1-18.el7.x86_64.rpm contains /usr/bin/wrap_j2k_in_mj2 contains glibc 2.15
INFO     cve_bin_tool.VersionScanner:version_scanner.py:162 /tmp/package_test-fww1_tm6/openjpeg-1.5.1-18.el7.x86_64.rpm contains /usr/bin/image_to_j2k contains glibc 2.15
INFO     cve_bin_tool.VersionScanner:version_scanner.py:162 /tmp/package_test-fww1_tm6/openjpeg-1.5.1-18.el7.x86_64.rpm contains /usr/bin/frames_to_mj2 contains glibc 2.15
INFO     cve_bin_tool.VersionScanner:version_scanner.py:162 /tmp/package_test-fww1_tm6/openjpeg-1.5.1-18.el7.x86_64.rpm contains /usr/bin/j2k_to_image contains glibc 2.15

----------- coverage: platform linux, python 3.8.6-final-0 -----------
```

if you look there, you can see that it's finding glibc in the rpm you're using as test data, but it is NOT finding openjpeg at all.  That means one of two things:

1. Your signature is bad
2. Your test data is bad

In this case, i believe it is the former: the signature you specified couldn't be found in any of the test files, and thus is is probably not a good signature.  When you were figuring out that signature, what file did you find it in?  I'm guessing from the research @pdxjohnny did, the answer is "you never found that string anywhere" so the fact that the checker can't find it is kind of unsurprising.  

I'm guessing that at your current skill level, you should just put a note in #1019  saying "I wasn't able to find a signature for this" and move on to a checker for something with an easier-to-find signature.  there's plenty of things that don't work  easily, and it looks like this is one of them.

pdxjohnny commented:
Reposting my gitter response here for posterity

> Essentially you need to write a `get_version()` method for the checker class you created that will return the version number of the library. How you implement it is up to you. I suggest checking to see if the library name is in the lines passed to the method and then look for the `@1.5.1` or whatever. Maybe maintain a list of version that it might be, maybe we can cross check with the cve database to dump all the version numbers that are vulnerable, not sure, start with the first part then maybe move onto the list of known versions.

chaitanyamogal commented:
@terriko @pdxjohnny  Please review it passes all tests.

terriko reviewed (commented):
well, filename based signatures aren't my favourite (since they often only work on fedora/centos/redhat) but this seems like the best we can find.

One question, though, is there some reason you didn't include the other filenames used by openjpeg?  In my test, it looks like they also have the same signature:

```console
(venv3.8) [terri@cedar tmp]$ for a in usr/bin/*; do echo "######### $a"; strings $a | grep openjpeg; done
######### usr/bin/extract_j2k_from_mj2
/builddir/build/BUILD/openjpeg-1.5.1/applications/mj2/extract_j2k_from_mj2.c
######### usr/bin/frames_to_mj2
######### usr/bin/image_to_j2k
libopenjpeg.so.1
/builddir/build/BUILD/openjpeg-1.5.1/applications/codec/convert.c
@/builddir/build/BUILD/openjpeg-1.5.1/applications/common/color.c
######### usr/bin/j2k_dump
libopenjpeg.so.1
/builddir/build/BUILD/openjpeg-1.5.1/applications/codec/convert.c
@/builddir/build/BUILD/openjpeg-1.5.1/applications/common/color.c
######### usr/bin/j2k_to_image
libopenjpeg.so.1
/builddir/build/BUILD/openjpeg-1.5.1/applications/codec/convert.c
@/builddir/build/BUILD/openjpeg-1.5.1/applications/common/color.c
######### usr/bin/mj2_to_frames
/builddir/build/BUILD/openjpeg-1.5.1/applications/mj2/mj2_to_frames.c
?/builddir/build/BUILD/openjpeg-1.5.1/applications/common/color.c
######### usr/bin/wrap_j2k_in_mj2
/builddir/build/BUILD/openjpeg-1.5.1/applications/mj2/wrap_j2k_in_mj2.c
````

(Sorry about my obsessive one line bash scripting, but that's basically printing the name of each file and then searching for strings in it.  As you can see, only `frames_to_mj2` doesn't seem to have the string included.)

chaitanyamogal commented:
hey, @terriko as you mention I have added all `FILENAME_PATTERNS` present. I think this is the best we can find in it.

terriko reviewed (approved):
Okay, I think this is the best we can do.  It may only work for some packages, but that's better than nothing/
